File: backend_fastapi/app/core/vault.py
import base64
import json
import os
from typing import Any

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _derive_key(secret: str) -> bytes:
    import hashlib
    return hashlib.scrypt(secret.encode("utf-8"), salt=b"agentpermit-vault-v1", n=2**14, r=8, p=1, dklen=32)


def encrypt_credential(value: dict[str, Any], master_secret: str) -> dict[str, Any]:
    key = _derive_key(master_secret)
    aes = AESGCM(key)
    iv = os.urandom(12)
    data = json.dumps(value, separators=(",", ":"), sort_keys=True).encode("utf-8")
    ct = aes.encrypt(iv, data, None)
    return {
        "version": 1,
        "algorithm": "aes-256-gcm",
        "iv": base64.b64encode(iv).decode(),
        "ciphertext": base64.b64encode(ct).decode(),
    }


def decrypt_credential(record: dict[str, Any], master_secret: str) -> dict[str, Any]:
    if record.get("version") != 1:
        raise ValueError("Unsupported credential record")
    key = _derive_key(master_secret)
    aes = AESGCM(key)
    iv = base64.b64decode(record["iv"])
    ct = base64.b64decode(record["ciphertext"])
    pt = aes.decrypt(iv, ct, None)
    return json.loads(pt.decode("utf-8"))


def redact_credential(item: dict[str, Any]) -> dict[str, Any]:
    safe = {k: v for k, v in item.items() if k != "encrypted"}
    safe["configured"] = bool(item.get("encrypted"))
    return safe

File: packages/core/vault.mjs
import crypto from 'node:crypto';

function deriveKey(secret) {
  return crypto.scryptSync(String(secret), 'agentpermit-vault-v1', 32);
}

export function encryptCredential(value, secret) {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', deriveKey(secret), iv);
  const plaintext = Buffer.from(JSON.stringify(value));
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  return {
    version: 1,
    algorithm: 'aes-256-gcm',
    iv: iv.toString('base64'),
    tag: cipher.getAuthTag().toString('base64'),
    ciphertext: ciphertext.toString('base64')
  };
}

export function decryptCredential(record, secret) {
  if (!record || record.version !== 1) throw new Error('Unsupported credential record');
  const decipher = crypto.createDecipheriv('aes-256-gcm', deriveKey(secret), Buffer.from(record.iv, 'base64'));
  decipher.setAuthTag(Buffer.from(record.tag, 'base64'));
  const plaintext = Buffer.concat([decipher.update(Buffer.from(record.ciphertext, 'base64')), decipher.final()]);
  return JSON.parse(plaintext.toString('utf8'));
}

export function redactCredential(item) {
  const { encrypted, ...safe } = item;
  return { ...safe, configured: Boolean(encrypted) };
}
